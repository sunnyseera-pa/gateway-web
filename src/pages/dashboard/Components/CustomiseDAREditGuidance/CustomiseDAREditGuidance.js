import { convertFromRaw, convertToRaw, EditorState } from 'draft-js';
import { has } from 'lodash';
import { draftToMarkdown, markdownToDraft } from 'markdown-draft-js';
import React, { useState } from 'react';
import { Button, Modal } from 'react-bootstrap';
import 'react-draft-wysiwyg/dist/react-draft-wysiwyg.css';
import { useTranslation } from 'react-i18next';
import { NotificationManager } from 'react-notifications';
import publishersService from '../../../../services/publishers';
import { WysiwygEditor } from '../../../commonComponents/WysiwygEditor/WysiwygEditor';
import './CustomiseDAREditGuidance.scss';

const baseURL = require('../../../commonComponents/BaseURL').getURL();

export const EditHowToRequestAccessPage = ({ show, onHide, publisherDetails }) => {
    const body = has(publisherDetails, 'dataRequestModalContent.body') ? publisherDetails.dataRequestModalContent.body : '';
    const [contentState] = useState(convertFromRaw(markdownToDraft(body)));
    const [editorState, setEditorState] = useState(EditorState.createWithContent(contentState));
    const [showConfirm, setShowConfirm] = useState(false);
    const [showCancel, setShowCancel] = useState(false);

    const { t } = useTranslation();

    const publishersRequest = publishersService.usePatchModalContent(null, {
        onError: ({ title, message }) => {
            NotificationManager.error(message, title, 10000);
        },
    });

    const handleConfirmOk = React.useCallback(async () => {
        const content = draftToMarkdown(convertToRaw(editorState.getCurrentContent()));

        const { _id } = publisherDetails;

        publishersRequest
            .mutateAsync({
                _id,
                content,
            })
            .then(() => {
                window.location.href = '/account?tab=customisedataaccessrequests_guidance&publishedDARContent=true';
            });
    }, [publisherDetails._id]);

    const handlePublish = React.useCallback(() => {
        setShowConfirm(true);
    }, []);

    const handleConfirmCancel = React.useCallback(() => {
        setShowConfirm(false);
    }, []);

    const handleCancel = React.useCallback(() => {
        setShowCancel(true);
    }, []);

    const handleCancelCancel = React.useCallback(() => {
        setShowCancel(false);
    }, []);

    const handleCancelOk = React.useCallback(() => {
        onHide();
    }, []);

    let modalContent = {
        header: `Edit 'Applicant guidance for requesting access to data'`,
        body: `The guidance below will be displayed to all data applicants at the beginning of their access journey. To ensure that applicants are prepared for the process, include all necessary information such as; what to do before they submit an application, the cost of accessing data, and any other resources that data applicants would find useful.`,
        cancel: (
            <Button variant='medium' className='cancelButton dark-14 mr-2' onClick={handleCancel}>
                {t('buttons.cancel')}
            </Button>
        ),
        confirm: (
            <Button
                data-test-id='add-collection-publish'
                variant='primary'
                className='publishButton white-14-semibold'
                type='submit'
                onClick={handlePublish}>
                {t('buttons.publish')}
            </Button>
        ),
    };

    if (showConfirm) {
        modalContent = {
            header: `Publish 'Applicant guidance for requesting access to data'`,
            body: `Are you sure you want to publish your updates to the 'Applicant guidance for requesting access to data'?`,
            cancel: (
                <Button variant='medium' className='cancelButton dark-14 mr-2' onClick={handleConfirmCancel}>
                    {t('buttons.neverMind')}
                </Button>
            ),
            confirm: (
                <Button
                    data-test-id='add-collection-publish'
                    variant='primary'
                    className='publishButton white-14-semibold'
                    type='submit'
                    onClick={handleConfirmOk}>
                    {t('buttons.confirm')}
                </Button>
            ),
        };
    } else if (showCancel) {
        modalContent = {
            header: `Cancel 'Applicant guidance for requesting access to data'`,
            body: `Are you sure you want to leave this page? Any changes you have made have not been published and will not be saved.`,
            cancel: (
                <Button variant='medium' className='cancelButton dark-14 mr-2' onClick={handleCancelCancel}>
                    {t('buttons.neverMind')}
                </Button>
            ),
            confirm: (
                <Button
                    data-test-id='add-collection-publish'
                    variant='primary'
                    className='publishButton white-14-semibold'
                    type='submit'
                    onClick={handleCancelOk}>
                    {t('buttons.cancel')}
                </Button>
            ),
        };
    }

    return (
        <Modal show={show} onHide={onHide} className={showConfirm || showCancel ? 'modal-md' : 'modal-xl'}>
            <Modal.Header>
                <h1 className='black-20-semibold'>{modalContent.header}</h1>
            </Modal.Header>
            <Modal.Body>
                <p className='soft-black-14'>{modalContent.body}</p>
                {!showConfirm && !showCancel && (
                    <WysiwygEditor data-testid='wysiwyg-editor' editorState={editorState} onEditorStateChange={setEditorState} />
                )}
            </Modal.Body>
            <Modal.Footer>
                <div>{modalContent.cancel}</div>
                <div className={!showCancel && !showConfirm && 'd-flex justify-content-end flex-grow'}>{modalContent.confirm}</div>
            </Modal.Footer>
        </Modal>
    );
};

export default EditHowToRequestAccessPage;
