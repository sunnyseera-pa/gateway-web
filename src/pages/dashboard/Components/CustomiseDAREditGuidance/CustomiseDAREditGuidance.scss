.cancelButton {
    min-width: 125px;
}
