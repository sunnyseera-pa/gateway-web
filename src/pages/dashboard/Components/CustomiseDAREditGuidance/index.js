import CustomiseDAREditGuidance from './CustomiseDAREditGuidance';

export default CustomiseDAREditGuidance;
