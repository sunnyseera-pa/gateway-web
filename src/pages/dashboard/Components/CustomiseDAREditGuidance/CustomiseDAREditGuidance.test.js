import { render, within, waitFor, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import React from 'react';
import CustomiseDAREditGuidance from './CustomiseDAREditGuidance';

let containerDiv;
const mockProps = {
    show: true,
    onHide: jest.fn(),
    publisherDetails: {
        dataRequestModalContent: {
            header: '',
            body: 'test content',
            footer: '',
        },
        publisherDetails: {
            accessRights: [],
            dataUseLimitation: [],
            dataUseRequirements: [],
            name: 'CUTOMISE DAR FORM',
            memberOf: 'OTHER',
            contactPoint: '',
        },
        active: true,
        allowsMessaging: false,
        workflowEnabled: false,
        allowAccessRequestManagement: true,
        uses5Safes: false,
        _id: '626a577d9d8825866696b1b3',
        name: 'OTHER > CUTOMISE DAR FORM',
        mdcFolderId: '1170da5d-4310-4379-92a4-56d557e8e45c',
        dataRequestModalContentUpdatedBy: 2623650828181272,
        dataRequestModalContentUpdatedOn: '2022-04-28T11:16:27.740Z',
    },
};

describe('Given the CustomiseDAREditGuidance component', () => {
    describe('When it is rendered', () => {
        let wrapper;

        beforeAll(() => {
            wrapper = render(<CustomiseDAREditGuidance {...mockProps} />, {
                wrapper: Providers,
            });
        });

        it('Should match the snapshot', async () => {
            expect(containerDiv).toMatchSnapshot();
        });

        it('Then the title and description should be rendered', async () => {
            expect(wrapper.getByTestId('modalHeading')).toHaveTextContent('Edit your guidance to data applicants');
            expect(wrapper.getByTestId('modalDesc')).toHaveTextContent(
                "The text below is shown to data applicants when they click on the 'How to access' button for CUTOMISE DAR FORM datasets"
            );
        });

        describe('publish modal', () => {
            it('Then onClick Publish Button the Modal should open', async () => {
                fireEvent.click(wrapper.getByTestId('publish-guidance'));
                await waitFor(() =>
                    expect(wrapper.getByTestId('modalHeading')).toHaveTextContent(
                        "Publish 'Applicant guidance for requesting access to data'"
                    )
                );
                await waitFor(() =>
                    expect(wrapper.getByTestId('modalDesc')).toHaveTextContent(
                        'Selecting Publish will make your updates live on the Gateway for CUTOMISE DAR FORM datasets.'
                    )
                );
            });
            it('Then onClick Cancel Button the Modal should be closed and previous modal should be shown', async () => {
                fireEvent.click(wrapper.getByTestId('confirm-cancel-publish'));
                await waitFor(() => expect(wrapper.getByTestId('modalHeading')).toHaveTextContent('Edit your guidance to data applicants'));
            });
        });

        describe('cancel publish modal', () => {
            it('Then onClick Cancel Button the cancel Modal should open', async () => {
                fireEvent.click(wrapper.getByTestId('cancel-publish'));
                await waitFor(() =>
                    expect(wrapper.getByTestId('modalHeading')).toHaveTextContent(
                        "Cancel 'Applicant guidance for requesting access to data'"
                    )
                );
                await waitFor(() =>
                    expect(wrapper.getByTestId('modalDesc')).toHaveTextContent(
                        'Are you sure you want to leave this page? Any changes you have made have not been published and will not be saved.'
                    )
                );
            });
            it('Then onClick Cancel Button the Modal should be closed and previous modal should be shown', async () => {
                fireEvent.click(wrapper.getByTestId('confirm-cancel'));
                await waitFor(() => expect(wrapper.getByTestId('modalHeading')).toHaveTextContent('Edit your guidance to data applicants'));
            });
        });
    });
});
