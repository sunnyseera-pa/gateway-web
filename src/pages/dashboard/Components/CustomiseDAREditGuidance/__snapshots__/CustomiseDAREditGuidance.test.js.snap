// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the CustomiseDAREditGuidance component When it is rendered Should match the snapshot 1`] = `undefined`;
